src/server/tiered_storage.cc

```diff
@@ -355,6 +356,68 @@ void TieredStorage::Free(PrimeIterator it, DbTableStats* stats) {
   stats->tiered_size -= len;
 }
 
+size_t TieredStorage::GetBinSize(size_t blob_len) {
+  CHECK(blob_len <= kMaxSmallBin);
+  unsigned bin_index = SmallToBin(blob_len);
+  return kSmallBins[bin_index];
+}
+
+void TieredStorage::Defrag(DbIndex db_index) {
+  // start scanning from a random position in the ref_cnt_ table.
+  auto refcnt_it = std::next(std::begin(page_refcnt_), rand() % page_refcnt_.size());
+
+  for (unsigned int i = 0; i < num_pages_to_scan_; ++i) {
+    // check the utilization of the page.
+    size_t bin_size = refcnt_it->second.second;
+    unsigned max_entries = NumEntriesInSmallBin(bin_size);
+    float bin_util = (float)(refcnt_it->second.first) / (float)max_entries;
+
+    // if bin is underutilized, start defragmentation.
+    if (bin_util < defrag_bin_util_threshold_) {
+      // 1. retrieve all the hash values stored in this page.
+      size_t hash_section_len = max_entries * kBytesPerHash;
+      std::vector<char> page(kBlockLen);
+      uint32_t offs_page = refcnt_it->first;
+      auto ec = Read(offs_page * kBlockLen, kBlockLen, &page[0]);
+      CHECK(!ec) << "Error when reading a page to be defragmented!";
+
+      for (unsigned int j = 0; j < max_entries; ++j) {
+        uint64_t key_hash = absl::little_endian::Load64(&page[j * kBytesPerHash]);
+        auto prime_it = db_slice_.GetDBTable(db_index)->prime.FindByHash(key_hash);
+
+        // if the key still exists, load the key into memory and reschedule offload
+        if (!prime_it.is_done()) {
+          PrimeValue* entry = &prime_it->second;
+          auto [offset, len] = entry->GetExternalSlice();
+          if (entry->IsExternal()) {
+            std::string value(page.data() + hash_section_len + j * bin_size, len);
+            auto* stats = db_slice_.MutableStats(db_index);
+            entry->SetString(value);
+            size_t heap_size = entry->MallocUsed();
+            stats->AddTypeMemoryUsage(entry->ObjType(), heap_size);
+            ScheduleOffload(db_index, prime_it);
+
+            bool has_expire = entry->HasExpire();
+            entry->Reset();
+            entry->SetExpire(has_expire);
+
+            stats->tiered_entries -= 1;
+            stats->tiered_size -= len;
+          }
+        }
+      }
+      // remove this entry from page_refcnt_
+      page_refcnt_.erase(refcnt_it);
+      alloc_.Free(offs_page * kBlockLen, kBlockLen);
```

Comment:
how do you continue in the scaning if you free a page? I dont see how this flow is handles as you erase the it from page_refcnt_ and then you continue in the for loop which will access the deleted iterator

Reply:
> how do you continue in the scaning if you free a page? I dont see how this flow is handles as you erase the it from page_refcnt_ and then you continue in the for loop which will access the deleted iterator

i did a revision, please see if the new change makes sense.